core/backend/src/SheetIndex.ts

```diff
@@ -0,0 +1,326 @@
+    sheetIndexModel: Id64String,
+    parentId: Id64String,
+    name: string,
+    priority: number,
+  ) {
+    const props: SheetIndexFolderProps = this.createProps(iModelDb, sheetIndexModel, parentId, name, priority);
+    return new this(props, iModelDb);
+  }
+
+  /** Create a new SheetIndexFolder
+   * @param iModelDb The iModel
+   * @param sheetIndexModel The [[SheetIndexModel]]
+   * @param parentId The [[SheetIndex]] or [[SheetIndexFolder]] that is parent to this Entry
+   * @param name The name of the SheetIndexFolder
+   * @param priority The priority of the SheetIndexFolder
+   * @returns The Id of the newly inserted SheetIndexFolder element.
+   * @throws [[IModelError]] if unable to create the element.
+   */
+  public static insert(
+    iModelDb: IModelDb,
+    sheetIndexModel: Id64String,
+    parentId: Id64String,
+    name: string,
+    priority: number,
+  ): Id64String {
+    const instance = this.create(iModelDb, sheetIndexModel, parentId, name, priority);
+    const elements = iModelDb.elements;
+    instance.id = elements.insertElement(instance.toJSON());
+    return instance.id;
+  }
+}
+
+/** A SheetIndexReference used to include a [[SheetIndex]] hierarchy into another one.
+ * @beta
+*/
+export class SheetIndexReference extends SheetIndexEntry {
+  public static override get className(): string { return "SheetIndexReference"; }
+  /** The bis:SheetIndex that this bis:SheetIndexReference is pointing to. */
+  public sheetIndex?: SheetIndexReferenceRefersToSheetIndex;
+  protected constructor(props: SheetIndexReferenceProps, iModel: IModelDb) {
+    super(props, iModel);
+    if (props.sheetIndex) {
+      const sheetIndex = iModel.elements.tryGetElement<SheetIndex>(props.sheetIndex.id);
+      if (!sheetIndex)
+        throw new IModelError(IModelStatus.NotFound, "SheetIndex not found");
+
+      this.sheetIndex = new SheetIndexReferenceRefersToSheetIndex(props.sheetIndex.id);
+    }
+  }
+
+  protected static createReferenceRelationshipProps(id: Id64String): RelatedElementProps {
+    return { id, relClassName: SheetIndexReferenceRefersToSheetIndex.classFullName };
+  }
+
+  public override toJSON(): SheetIndexReferenceProps { // This override only specializes the return type
+    return {
+      ...super.toJSON(),
+      sheetIndex: this.sheetIndex ? this.sheetIndex.toJSON() : undefined,
+    } as SheetIndexReferenceProps;
+  }
+
+  /** Create a new SheetIndexReference
+   * @param iModelDb The iModel
+   * @param sheetIndexModel The [[SheetIndexModel]]
+   * @param parentId The [[SheetIndex]] or [[SheetIndexFolder]] that is parent to this Entry
+   * @param name The name of the SheetIndexReference
+   * @param priority The priority of the SheetIndexReference
+   * @param sheetIndexId The Sheet Index referenced by the SheetIndexReference
+   * @returns The newly constructed SheetIndexReference element.
+   * @throws [[IModelError]] if unable to create the element.
+   */
+  public static create(
```

Comment:
Too many args here and in `insert`, same for `SheetIndexFolder` methods. Consolidate into an "arguments" object so people can't screw up the order of the 4 string arguments.

Reply:
Refactored